File: zig_compiler/runtime/math_ops.c
//
// math_ops.c
// FasterBASIC QBE Runtime Library - Math Operations
//
// This file implements mathematical functions for BASIC programs.
//

#include "basic_runtime.h"
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

// =============================================================================
// Absolute Value
// =============================================================================

int32_t basic_abs_int(int32_t x) {
    return x < 0 ? -x : x;
}

double basic_abs_double(double x) {
    return fabs(x);
}

// =============================================================================
// Square Root
// =============================================================================

double basic_sqrt(double x) {
    if (x < 0.0) {
        basic_error_msg("Square root of negative number");
        return 0.0;
    }
    return sqrt(x);
}

// =============================================================================
// Power
// =============================================================================

double basic_pow(double base, double exponent) {
    // Handle special cases
    if (base == 0.0 && exponent < 0.0) {
        basic_error_msg("Division by zero in power operation");
        return 0.0;
    }
    
    return pow(base, exponent);
}

// =============================================================================
// Extended Exponentials and Logarithms
// =============================================================================

double basic_exp2(double x) {
    return exp2(x);
}

double basic_expm1(double x) {
    return expm1(x);
}

double basic_log10(double x) {
    if (x <= 0.0) {
        basic_error_msg("Logarithm base 10 of non-positive number");
        return 0.0;
    }
    return log10(x);
}

double basic_log1p(double x) {
    if (x <= -1.0) {
        basic_error_msg("Logarithm of 1 + x with x <= -1");
        return 0.0;
    }
    return log1p(x);
}

// =============================================================================
// Trigonometric Functions
// =============================================================================

double basic_sin(double x) {
    return sin(x);
}

double basic_cos(double x) {
    return cos(x);
}

double basic_tan(double x) {
    return tan(x);
}

double basic_asin(double x) {
    if (x < -1.0 || x > 1.0) {
        basic_error_msg("ASIN domain error");
        return 0.0;
    }
    return asin(x);
}

double basic_acos(double x) {
    if (x < -1.0 || x > 1.0) {
        basic_error_msg("ACOS domain error");
        return 0.0;
    }
    return acos(x);
}

double basic_atan(double x) {
    return atan(x);
}

double basic_atan2(double y, double x) {
    return atan2(y, x);
}

// Hyperbolic functions
double basic_sinh(double x) {
    return sinh(x);
}

double basic_cosh(double x) {
    return cosh(x);
}

double basic_tanh(double x) {
    return tanh(x);
}

double basic_asinh(double x) {
    return asinh(x);
}

double basic_acosh(double x) {
    if (x < 1.0) {
        basic_error_msg("ACOSH domain error");
        return 0.0;
    }
    return acosh(x);
}

double basic_atanh(double x) {
    if (x <= -1.0 || x >= 1.0) {
        basic_error_msg("ATANH domain error");
        return 0.0;
    }
    return atanh(x);
}

// =============================================================================
// Logarithm and Exponential
// =============================================================================

double basic_log(double x) {
    if (x <= 0.0) {
        basic_error_msg("Logarithm of non-positive number");
        return 0.0;
    }
    return log(x);
}

double basic_exp(double x) {
    return exp(x);
}

// =============================================================================
// Power Helpers and Roots
// =============================================================================

double basic_cbrt(double x) {
    return cbrt(x);
}

double basic_hypot(double x, double y) {
    return hypot(x, y);
}

double basic_fmod(double x, double y) {
    if (y == 0.0) {
        basic_error_msg("FMOD division by zero");
        return 0.0;
    }
    return fmod(x, y);
}

double basic_remainder(double x, double y) {
    if (y == 0.0) {
        basic_error_msg("REMAINDER division by zero");
        return 0.0;
    }
    return remainder(x, y);
}

double basic_floor(double x) {
    return floor(x);
}

double basic_ceil(double x) {
    return ceil(x);
}

double basic_trunc(double x) {
    return trunc(x);
}

double basic_round(double x) {
    return round(x);
}

double basic_copysign(double x, double y) {
    return copysign(x, y);
}

double basic_erf(double x) {
    return erf(x);
}

double basic_erfc(double x) {
    return erfc(x);
}

double basic_tgamma(double x) {
    return tgamma(x);
}

double basic_lgamma(double x) {
    return lgamma(x);
}

double basic_nextafter(double x, double y) {
    return nextafter(x, y);
}

double basic_fmax(double x, double y) {
    return fmax(x, y);
}

double basic_fmin(double x, double y) {
    return fmin(x, y);
}

double basic_fma(double x, double y, double z) {
    return fma(x, y, z);
}

double basic_deg(double radians) {
    return radians * (180.0 / M_PI);
}

double basic_rad(double degrees) {
    return degrees * (M_PI / 180.0);
}

double basic_sigmoid(double x) {
    return 1.0 / (1.0 + exp(-x));
}

double basic_logit(double x) {
    if (x <= 0.0 || x >= 1.0) {
        basic_error_msg("LOGIT domain error (0<x<1)");
        return 0.0;
    }
    return log(x / (1.0 - x));
}

double basic_normpdf(double x) {
    const double inv_sqrt_2pi = 0.3989422804014327; // 1/sqrt(2*pi)
    return inv_sqrt_2pi * exp(-0.5 * x * x);
}

double basic_normcdf(double x) {
    // Standard normal CDF via erf
    return 0.5 * (1.0 + erf(x / M_SQRT2));
}

double basic_fact(double n) {
    if (n < 0.0) {
        basic_error_msg("FACTORIAL of negative number");
        return 0.0;
    }
    if (n > 170.0) {
        basic_error_msg("FACTORIAL overflow");
        return 0.0;
    }
    // Use gamma(n+1)
    return tgamma(n + 1.0);
}

double basic_comb(double n, double k) {
    if (k < 0.0 || n < 0.0 || k > n) {
        basic_error_msg("COMB domain error");
        return 0.0;
    }
    return floor(0.5 + exp(lgamma(n + 1.0) - lgamma(k + 1.0) - lgamma(n - k + 1.0)));
}

double basic_perm(double n, double k) {
    if (k < 0.0 || n < 0.0 || k > n) {
        basic_error_msg("PERM domain error");
        return 0.0;
    }
    return floor(0.5 + exp(lgamma(n + 1.0) - lgamma(n - k + 1.0)));
}

double basic_clamp(double x, double minv, double maxv) {
    if (minv > maxv) {
        double tmp = minv; minv = maxv; maxv = tmp;
    }
    if (x < minv) return minv;
    if (x > maxv) return maxv;
    return x;
}

double basic_lerp(double a, double b, double t) {
    return a + (b - a) * t;
}

double basic_pmt(double rate, double nper, double pv) {
    if (nper <= 0.0) {
        basic_error_msg("PMT nper must be > 0");
        return 0.0;
    }
    if (fabs(rate) < 1e-12) {
        return -(pv) / nper;
    }
    double r1 = pow(1.0 + rate, nper);
    return -(pv * rate * r1) / (r1 - 1.0);
}

double basic_pv(double rate, double nper, double pmt) {
    if (nper <= 0.0) {
        basic_error_msg("PV nper must be > 0");
        return 0.0;
    }
    if (fabs(rate) < 1e-12) {
        return -(pmt) * nper;
    }
    double r1 = pow(1.0 + rate, nper);
    return -pmt * (r1 - 1.0) / (rate * r1);
}

double basic_fv(double rate, double nper, double pmt) {
    if (nper <= 0.0) {
        basic_error_msg("FV nper must be > 0");
        return 0.0;
    }
    if (fabs(rate) < 1e-12) {
        return -(pmt) * nper;
    }
    double r1 = pow(1.0 + rate, nper);
    return -pmt * (r1 - 1.0) / rate;
}

// =============================================================================
// Random Number Generation
// =============================================================================

static bool rng_initialized = false;

double basic_rnd(void) {
    if (!rng_initialized) {
        srand((unsigned int)time(NULL));
        rng_initialized = true;
    }
    
    // Return random number between 0.0 and 1.0
    return (double)rand() / (double)RAND_MAX;
}

int32_t basic_rnd_int(int32_t min, int32_t max) {
    if (!rng_initialized) {
        srand((unsigned int)time(NULL));
        rng_initialized = true;
    }
    
    if (min > max) {
        int32_t temp = min;
        min = max;
        max = temp;
    }
    
    // Generate random integer in range [min, max]
    int32_t range = max - min + 1;
    return min + (rand() % range);
}

void basic_randomize(int32_t seed) {
    srand((unsigned int)seed);
    rng_initialized = true;
}

// Get random integer from 0 to n-1 (BASIC RAND function)
int32_t basic_rand(int32_t n) {
    if (n <= 0) return 0;
    return rand() % n;
}

// =============================================================================
// Integer and Sign Functions
// =============================================================================

int32_t basic_int(double x) {
    // INT() in BASIC truncates towards negative infinity (floor)
    return (int32_t)floor(x);
}

int32_t basic_sgn(double x) {
    if (x < 0.0) return -1;
    if (x > 0.0) return 1;
    return 0;
}

int32_t basic_fix(double x) {
    // FIX() in BASIC truncates towards zero
    return (int32_t)x;
}

int32_t math_cint(double x) {
    // CINT() in BASIC rounds to nearest integer
    return (int32_t)round(x);
}

File: zig_compiler/runtime/math_ops.zig
//
// math_ops.zig
// FasterBASIC Runtime — Math Operations
//
// Implements all mathematical functions for BASIC programs.
// Pure wrappers around libm with domain-error checking.
//

const std = @import("std");
const math = std.math;

// =========================================================================
// Extern declarations
// =========================================================================

extern fn basic_error_msg(msg: [*:0]const u8) void;

// C math library functions not directly in Zig std.math
extern fn fabs(x: f64) f64;
extern fn sqrt(x: f64) f64;
extern fn pow(base: f64, exp: f64) f64;
extern fn sin(x: f64) f64;
extern fn cos(x: f64) f64;
extern fn tan(x: f64) f64;
extern fn asin(x: f64) f64;
extern fn acos(x: f64) f64;
extern fn atan(x: f64) f64;
extern fn atan2(y: f64, x: f64) f64;
extern fn sinh(x: f64) f64;
extern fn cosh(x: f64) f64;
extern fn tanh(x: f64) f64;
extern fn asinh(x: f64) f64;
extern fn acosh(x: f64) f64;
extern fn atanh(x: f64) f64;
extern fn log(x: f64) f64;
extern fn exp(x: f64) f64;
extern fn exp2(x: f64) f64;
extern fn expm1(x: f64) f64;
extern fn log10(x: f64) f64;
extern fn log1p(x: f64) f64;
extern fn cbrt(x: f64) f64;
extern fn hypot(x: f64, y: f64) f64;
extern fn fmod(x: f64, y: f64) f64;
extern fn remainder(x: f64, y: f64) f64;
extern fn floor(x: f64) f64;
extern fn ceil(x: f64) f64;
extern fn trunc(x: f64) f64;
extern fn round(x: f64) f64;
extern fn copysign(x: f64, y: f64) f64;
extern fn erf(x: f64) f64;
extern fn erfc(x: f64) f64;
extern fn tgamma(x: f64) f64;
extern fn lgamma(x: f64) f64;
extern fn nextafter(x: f64, y: f64) f64;
extern fn fmax(x: f64, y: f64) f64;
extern fn fmin(x: f64, y: f64) f64;
extern fn fma(x: f64, y: f64, z: f64) f64;

// C stdlib for RNG
extern fn srand(seed: c_uint) void;
extern fn rand() c_int;
extern fn time(t: ?*anyopaque) c_long;

const RAND_MAX: c_int = 0x7fffffff;

const M_PI: f64 = 3.14159265358979323846;
const M_SQRT2: f64 = 1.41421356237309504880;

// =========================================================================
// Absolute Value
// =========================================================================

export fn basic_abs_int(x: i32) callconv(.c) i32 {
    return if (x < 0) -x else x;
}

export fn basic_abs_double(x: f64) callconv(.c) f64 {
    return fabs(x);
}

// =========================================================================
// Square Root
// =========================================================================

export fn basic_sqrt(x: f64) callconv(.c) f64 {
    if (x < 0.0) {
        basic_error_msg("Square root of negative number");
        return 0.0;
    }
    return sqrt(x);
}

// =========================================================================
// Power
// =========================================================================

export fn basic_pow(base_val: f64, exponent: f64) callconv(.c) f64 {
    if (base_val == 0.0 and exponent < 0.0) {
        basic_error_msg("Division by zero in power operation");
        return 0.0;
    }
    return pow(base_val, exponent);
}

// =========================================================================
// Extended Exponentials and Logarithms
// =========================================================================

export fn basic_exp2(x: f64) callconv(.c) f64 {
    return exp2(x);
}

export fn basic_expm1(x: f64) callconv(.c) f64 {
    return expm1(x);
}

export fn basic_log10(x: f64) callconv(.c) f64 {
    if (x <= 0.0) {
        basic_error_msg("Logarithm base 10 of non-positive number");
        return 0.0;
    }
    return log10(x);
}

export fn basic_log1p(x: f64) callconv(.c) f64 {
    if (x <= -1.0) {
        basic_error_msg("Logarithm of 1 + x with x <= -1");
        return 0.0;
    }
    return log1p(x);
}

// =========================================================================
// Trigonometric Functions
// =========================================================================

export fn basic_sin(x: f64) callconv(.c) f64 {
    return sin(x);
}

export fn basic_cos(x: f64) callconv(.c) f64 {
    return cos(x);
}

export fn basic_tan(x: f64) callconv(.c) f64 {
    return tan(x);
}

export fn basic_asin(x: f64) callconv(.c) f64 {
    if (x < -1.0 or x > 1.0) {
        basic_error_msg("ASIN domain error");
        return 0.0;
    }
    return asin(x);
}

export fn basic_acos(x: f64) callconv(.c) f64 {
    if (x < -1.0 or x > 1.0) {
        basic_error_msg("ACOS domain error");
        return 0.0;
    }
    return acos(x);
}

export fn basic_atan(x: f64) callconv(.c) f64 {
    return atan(x);
}

export fn basic_atan2(y: f64, x: f64) callconv(.c) f64 {
    return atan2(y, x);
}

// Hyperbolic functions
export fn basic_sinh(x: f64) callconv(.c) f64 {
    return sinh(x);
}

export fn basic_cosh(x: f64) callconv(.c) f64 {
    return cosh(x);
}

export fn basic_tanh(x: f64) callconv(.c) f64 {
    return tanh(x);
}

export fn basic_asinh(x: f64) callconv(.c) f64 {
    return asinh(x);
}

export fn basic_acosh(x: f64) callconv(.c) f64 {
    if (x < 1.0) {
        basic_error_msg("ACOSH domain error");
        return 0.0;
    }
    return acosh(x);
}

export fn basic_atanh(x: f64) callconv(.c) f64 {
    if (x <= -1.0 or x >= 1.0) {
        basic_error_msg("ATANH domain error");
        return 0.0;
    }
    return atanh(x);
}

// =========================================================================
// Logarithm and Exponential
// =========================================================================

export fn basic_log(x: f64) callconv(.c) f64 {
    if (x <= 0.0) {
        basic_error_msg("Logarithm of non-positive number");
        return 0.0;
    }
    return log(x);
}

export fn basic_exp(x: f64) callconv(.c) f64 {
    return exp(x);
}

// =========================================================================
// Power Helpers and Roots
// =========================================================================

export fn basic_cbrt(x: f64) callconv(.c) f64 {
    return cbrt(x);
}

export fn basic_hypot(x: f64, y: f64) callconv(.c) f64 {
    return hypot(x, y);
}

export fn basic_fmod(x: f64, y: f64) callconv(.c) f64 {
    if (y == 0.0) {
        basic_error_msg("FMOD division by zero");
        return 0.0;
    }
    return fmod(x, y);
}

export fn basic_remainder(x: f64, y: f64) callconv(.c) f64 {
    if (y == 0.0) {
        basic_error_msg("REMAINDER division by zero");
        return 0.0;
    }
    return remainder(x, y);
}

export fn basic_floor(x: f64) callconv(.c) f64 {
    return floor(x);
}

export fn basic_ceil(x: f64) callconv(.c) f64 {
    return ceil(x);
}

export fn basic_trunc(x: f64) callconv(.c) f64 {
    return trunc(x);
}

export fn basic_round(x: f64) callconv(.c) f64 {
    return round(x);
}

export fn basic_copysign(x: f64, y: f64) callconv(.c) f64 {
    return copysign(x, y);
}

export fn basic_erf(x: f64) callconv(.c) f64 {
    return erf(x);
}

export fn basic_erfc(x: f64) callconv(.c) f64 {
    return erfc(x);
}

export fn basic_tgamma(x: f64) callconv(.c) f64 {
    return tgamma(x);
}

export fn basic_lgamma(x: f64) callconv(.c) f64 {
    return lgamma(x);
}

export fn basic_nextafter(x: f64, y: f64) callconv(.c) f64 {
    return nextafter(x, y);
}

export fn basic_fmax(x: f64, y: f64) callconv(.c) f64 {
    return fmax(x, y);
}

export fn basic_fmin(x: f64, y: f64) callconv(.c) f64 {
    return fmin(x, y);
}

export fn basic_fma(x: f64, y: f64, z: f64) callconv(.c) f64 {
    return fma(x, y, z);
}

// =========================================================================
// Angle Conversion
// =========================================================================

export fn basic_deg(radians: f64) callconv(.c) f64 {
    return radians * (180.0 / M_PI);
}

export fn basic_rad(degrees: f64) callconv(.c) f64 {
    return degrees * (M_PI / 180.0);
}

// =========================================================================
// Statistical / Special Functions
// =========================================================================

export fn basic_sigmoid(x: f64) callconv(.c) f64 {
    return 1.0 / (1.0 + exp(-x));
}

export fn basic_logit(x: f64) callconv(.c) f64 {
    if (x <= 0.0 or x >= 1.0) {
        basic_error_msg("LOGIT domain error (0<x<1)");
        return 0.0;
    }
    return log(x / (1.0 - x));
}

export fn basic_normpdf(x: f64) callconv(.c) f64 {
    const inv_sqrt_2pi: f64 = 0.3989422804014327; // 1/sqrt(2*pi)
    return inv_sqrt_2pi * exp(-0.5 * x * x);
}

export fn basic_normcdf(x: f64) callconv(.c) f64 {
    return 0.5 * (1.0 + erf(x / M_SQRT2));
}

// =========================================================================
// Combinatorics
// =========================================================================

export fn basic_fact(n: f64) callconv(.c) f64 {
    if (n < 0.0) {
        basic_error_msg("FACTORIAL of negative number");
        return 0.0;
    }
    if (n > 170.0) {
        basic_error_msg("FACTORIAL overflow");
        return 0.0;
    }
    return tgamma(n + 1.0);
}

export fn basic_comb(n: f64, k: f64) callconv(.c) f64 {
    if (k < 0.0 or n < 0.0 or k > n) {
        basic_error_msg("COMB domain error");
        return 0.0;
    }
    return floor(0.5 + exp(lgamma(n + 1.0) - lgamma(k + 1.0) - lgamma(n - k + 1.0)));
}

export fn basic_perm(n: f64, k: f64) callconv(.c) f64 {
    if (k < 0.0 or n < 0.0 or k > n) {
        basic_error_msg("PERM domain error");
        return 0.0;
    }
    return floor(0.5 + exp(lgamma(n + 1.0) - lgamma(n - k + 1.0)));
}

// =========================================================================
// Clamping / Interpolation
// =========================================================================

export fn basic_clamp(x: f64, minv: f64, maxv: f64) callconv(.c) f64 {
    var lo = minv;
    var hi = maxv;
    if (lo > hi) {
        const tmp = lo;
        lo = hi;
        hi = tmp;
    }
    if (x < lo) return lo;
    if (x > hi) return hi;
    return x;
}

export fn basic_lerp(a: f64, b: f64, t: f64) callconv(.c) f64 {
    return a + (b - a) * t;
}

// =========================================================================
// Financial Functions
// =========================================================================

export fn basic_pmt(rate: f64, nper: f64, pv: f64) callconv(.c) f64 {
    if (nper <= 0.0) {
        basic_error_msg("PMT nper must be > 0");
        return 0.0;
    }
    if (fabs(rate) < 1e-12) {
        return -(pv) / nper;
    }
    const r1 = pow(1.0 + rate, nper);
    return -(pv * rate * r1) / (r1 - 1.0);
}

export fn basic_pv(rate: f64, nper: f64, pmt_val: f64) callconv(.c) f64 {
    if (nper <= 0.0) {
        basic_error_msg("PV nper must be > 0");
        return 0.0;
    }
    if (fabs(rate) < 1e-12) {
        return -(pmt_val) * nper;
    }
    const r1 = pow(1.0 + rate, nper);
    return -pmt_val * (r1 - 1.0) / (rate * r1);
}

export fn basic_fv(rate: f64, nper: f64, pmt_val: f64) callconv(.c) f64 {
    if (nper <= 0.0) {
        basic_error_msg("FV nper must be > 0");
        return 0.0;
    }
    if (fabs(rate) < 1e-12) {
        return -(pmt_val) * nper;
    }
    const r1 = pow(1.0 + rate, nper);
    return -pmt_val * (r1 - 1.0) / rate;
}

// =========================================================================
// Random Number Generation
// =========================================================================

var rng_initialized: bool = false;

export fn basic_rnd() callconv(.c) f64 {
    if (!rng_initialized) {
        srand(@intCast(time(null)));
        rng_initialized = true;
    }
    return @as(f64, @floatFromInt(rand())) / @as(f64, @floatFromInt(RAND_MAX));
}

export fn basic_rnd_int(min_val: i32, max_val: i32) callconv(.c) i32 {
    if (!rng_initialized) {
        srand(@intCast(time(null)));
        rng_initialized = true;
    }

    var lo = min_val;
    var hi = max_val;
    if (lo > hi) {
        const tmp = lo;
        lo = hi;
        hi = tmp;
    }

    const range: i32 = hi - lo + 1;
    return lo + @mod(rand(), range);
}

export fn basic_randomize(seed: i32) callconv(.c) void {
    srand(@intCast(seed));
    rng_initialized = true;
}

export fn basic_rand(n: i32) callconv(.c) i32 {
    if (n <= 0) return 0;
    return @mod(rand(), n);
}

// =========================================================================
// Integer and Sign Functions
// =========================================================================

export fn basic_int(x: f64) callconv(.c) i32 {
    // INT() in BASIC truncates towards negative infinity (floor)
    return @intFromFloat(floor(x));
}

export fn basic_sgn(x: f64) callconv(.c) i32 {
    if (x < 0.0) return -1;
    if (x > 0.0) return 1;
    return 0;
}

export fn basic_fix(x: f64) callconv(.c) i32 {
    // FIX() in BASIC truncates towards zero
    return @intFromFloat(x);
}

export fn math_cint(x: f64) callconv(.c) i32 {
    // CINT() in BASIC rounds to nearest integer
    return @intFromFloat(round(x));
}
